ipn/store/kubestore/store_kube.go

```diff
@@ -84,27 +116,79 @@ func (s *Store) ReadState(id ipn.StateKey) ([]byte, error) {
 
 // WriteState implements the StateStore interface.
 func (s *Store) WriteState(id ipn.StateKey, bs []byte) (err error) {
-	return s.updateStateSecret(map[string][]byte{string(id): bs})
+	defer func() {
+		if err == nil {
+			s.memory.WriteState(ipn.StateKey(sanitizeKey(id)), bs)
+		}
+	}()
+	return s.updateSecret(map[string][]byte{string(id): bs}, s.secretName)
 }
 
-// WriteTLSCertAndKey writes a TLS cert and key to domain.crt, domain.key fields of a Tailscale Kubernetes node's state
-// Secret.
-func (s *Store) WriteTLSCertAndKey(domain string, cert, key []byte) error {
-	return s.updateStateSecret(map[string][]byte{domain + ".crt": cert, domain + ".key": key})
+// WriteTLSCertAndKey writes a TLS cert and key to domain.crt, domain.key fields
+// of a Tailscale Kubernetes node's state Secret.
+func (s *Store) WriteTLSCertAndKey(domain string, cert, key []byte) (err error) {
+	defer func() {
+		if err == nil {
+			s.memory.WriteState(ipn.StateKey(sanitizeKey(domain+".crt")), cert)
+			s.memory.WriteState(ipn.StateKey(sanitizeKey(domain+".key")), key)
```

Comment:
If we get very unlucky, we could race with the periodic refresh goroutine. I think it's worth having a mutex around the whole of the `WriteTLSCertAndKey` and `ReadTLSCertAndKey` functions, and the relevant part of `loadCerts` just to make sure the Secret is always consistent with the in-memory store, and the cert is always consistent with the key.

Reply:
I think what I want to do is to have

```
func (s *Store) ReadStates(ids []ipn.StateKey) (map[ipn.StateKey][]byte, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
...
```
and similar for Write and call those from `WriteTLSCertAndKey`, `ReadTLSCertAndKey`.
I'm uneasy about doing a thing where we add a lock to the kube store given that we also lock the inner memory store- I think that could end up with deadlocks especially when the other `ReadState`, `WriteState` method will only lock the inner struct.

But then I'd rather put this Store interface change into a small separate PR

Reply:
I just added a TODO for now